core/src/main/java/jenkins/security/ApiTokenProperty.java

```diff
@@ -176,29 +317,169 @@ public String getDisplayName() {
          * the initial API token value. So we take the seed by hashing the secret + user ID.
          */
         public ApiTokenProperty newInstance(User user) {
-            return new ApiTokenProperty(API_KEY_SEED.mac(user.getId()));
+            if (!ApiTokenPropertyConfiguration.get().isTokenGenerationOnCreationEnabled()) {
+                // recommended way
+                return null;
+            }
+
+            return forceNewInstance(user, true);
+        }
+        
+        private ApiTokenProperty forceNewInstance(User user, boolean withLegacyToken) {
+            if(withLegacyToken){
+                return new ApiTokenProperty(API_KEY_SEED.mac(user.getId()));
+            }else{
+                return new ApiTokenProperty(null);
+            }
+        }
+    
+        // for Jelly view
+        @Restricted(NoExternalUse.class)
+        public boolean mustDisplayLegacyApiToken(User propertyOwner) {
+            ApiTokenProperty property = propertyOwner.getProperty(ApiTokenProperty.class);
+            if(property != null && property.apiToken != null){
+                return true;
+            }
+            return ApiTokenPropertyConfiguration.get().isCreationOfLegacyTokenEnabled();
+        }
+    
+        // for Jelly view
+        @Restricted(NoExternalUse.class)
+        public boolean hasCurrentUserRightToGenerateNewToken(User propertyOwner){
+            if (ADMIN_CAN_GENERATE_NEW_TOKENS && Jenkins.get().hasPermission(Jenkins.ADMINISTER)) {
+                return true;
+            }
+
+            User currentUser = User.current();
+            if (currentUser == null) {
+                // Anonymous
+                return false;
+            }
+
+            if (Jenkins.getAuthentication() == ACL.SYSTEM) {
+                // SYSTEM user is always eligible to see tokens
+                return true;
+            }
+
+            return User.idStrategy().equals(propertyOwner.getId(), currentUser.getId());
         }
 
+        /**
+         * @deprecated use {@link #doGenerateNewToken(User, StaplerResponse, String)} instead
+         */
+        @Deprecated
         @RequirePOST
         public HttpResponse doChangeToken(@AncestorInPath User u, StaplerResponse rsp) throws IOException {
+            // you are the user or you have ADMINISTER permission
+            u.checkPermission(Jenkins.ADMINISTER);
+
+            LOGGER.log(Level.WARNING, "Deprecated action /changeToken used, consider using /generateNewToken instead");
```

Comment:
This also happens when enabling legacy token options in the security configuration, and seems therefore inappropriate for a warning.